librad/src/net/gossip.rs

```diff
@@ -265,9 +271,10 @@ impl<S: Clone, R, W> Clone for Protocol<S, R, W> {
     }
 }
 
-impl<S, R, W> Protocol<S, R, W>
+impl<S, A, R, W> Protocol<S, A, R, W>
 where
-    S: LocalStorage + 'static,
+    S: LocalStorage<Update = A> + 'static,
+    for<'de> A: Serialize + Deserialize<'de> + Clone + Debug + Send + Sync + 'static,
```

Comment:
The `'static` lifetime doesn't seem right here

Reply:
Anyone's got an idea here? It seems to me like the lifetime of `A` should somehow be bound to the lifetime of `LocalStorage`, but I can't seem to figure out how. /cc @cloudhead @FintanH

Reply:
My lifetime-fu is not good enough to tell what's going on here. My _guess_ is that the `for<'de>` is the culprit and because it gets introduced the inference for `A` gets messed up.

What would it usually be inferred as? Would it not usually be `'static`? :thinking:

Reply:
Yeah so in my understanding the problem is as follows:
1. `A` is some arbitrary type with some bounds.
2. *iff* the concrete type implementing `A` contains references, then those references are going to need some kind of lifetime.
3. To "prove" to the compiler that it should not worry about any possible references inside of `A` and their lifetime, you have to say that references within `A` are valid for the whole program (`'static`). And this is the case because you are holding `A` in a struct with no lifetime parameters, so it must be `'static`.

Reply:
Nb. `'static` applied as a *bound* means that the type cannot have any references within it with a lifetime shorter than `'static`